packages/block-library/src/query/edit/query-block-setup.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { SelectControl, ToggleControl } from '@wordpress/components';
+import { createBlocksFromInnerBlocksTemplate } from '@wordpress/blocks';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import BlockSetup from './block-setup';
+
+const QueryBlockSetup = ( {
+	clientId,
+	attributes: { query },
+	setAttributes,
+	name,
+} ) => {
+	const { postType, inherit } = query;
+	const { replaceInnerBlocks } = useDispatch( blockEditorStore );
+	const { postTypes } = useSelect( ( select ) => {
+		const { getPostTypes } = select( 'core' );
+		const excludedPostTypes = [ 'attachment' ];
+		const filteredPostTypes = getPostTypes( { per_page: -1 } )?.filter(
+			( { viewable, slug } ) =>
+				viewable && ! excludedPostTypes.includes( slug )
+		);
+		return {
+			postTypes: filteredPostTypes,
+		};
+	}, [] );
+	const updateQuery = ( newQuery ) =>
+		setAttributes( { query: { ...query, ...newQuery } } );
+	const postTypesSelectOptions = useMemo(
+		() =>
+			( postTypes || [] ).map( ( { labels, slug } ) => ( {
+				label: labels.singular_name,
+				value: slug,
+			} ) ),
+		[ postTypes ]
+	);
+	const onFinish = ( innerBlocks ) => {
+		if ( innerBlocks ) {
+			replaceInnerBlocks(
+				clientId,
+				createBlocksFromInnerBlocksTemplate( innerBlocks ),
+				false
+			);
+		}
+	};
+	const blockSetupProps = {
```

Comment:
In React, passing down complex objects like that usually doesn't scale well because of re-rendering. You can achieve a similar goal using a composition of smaller components and hooks that abstract common functionalities.

Reply:
Yes! I'm waiting for the design and probably consolidate other blocks' use cases to refactor/change the code. We will definitely have more discussions about the implementation then :)